packages/session-replay-browser/src/events/events-idb-store.ts

```diff
@@ -0,0 +1,308 @@
+    try {
+      const tx = this.db?.transaction<'sessionCurrentSequence', 'readwrite'>(currentSequenceKey, 'readwrite');
+      if (!tx) {
+        return;
+      }
+      const sequenceEvents = await tx.store.get(sessionId);
+      if (!sequenceEvents) {
+        await tx.store.put({ sessionId, events: [event] });
+        return;
+      }
+      let eventsToSend;
+      if (this.shouldSplitEventsList(sequenceEvents.events, event)) {
+        eventsToSend = sequenceEvents.events;
+        // set store to empty array
+        await tx.store.put({ sessionId, events: [event] });
+      } else {
+        // add event to array
+        const updatedEvents = sequenceEvents.events.concat(event);
+        await tx.store.put({ sessionId, events: updatedEvents });
+      }
+
+      await tx.done;
+      if (!eventsToSend) {
+        return undefined;
+      }
+
+      const sequenceId = await this.storeSendingEvents(sessionId, eventsToSend);
+
+      if (!sequenceId) {
+        return undefined;
+      }
+
+      return {
+        events: eventsToSend,
+        sessionId,
+        sequenceId,
+      };
+    } catch (e) {
+      this.loggerProvider.warn(`${STORAGE_FAILURE}: ${e as string}`);
+    }
+    return undefined;
+  };
+
+  storeSendingEvents = async (sessionId: number, events: Events) => {
+    try {
+      const sequenceId = await this.db?.put<'sequencesToSend'>(sequencesToSendKey, {
+        sessionId: sessionId,
+        events: events,
+      });
+      return sequenceId;
+    } catch (e) {
+      this.loggerProvider.warn(`${STORAGE_FAILURE}: ${e as string}`);
+    }
+    return undefined;
+  };
+
+  cleanUpSessionEventsStore = async (sequenceId: number) => {
+    try {
+      await this.db?.delete<'sequencesToSend'>(sequencesToSendKey, sequenceId);
+    } catch (e) {
+      this.loggerProvider.warn(`${STORAGE_FAILURE}: ${e as string}`);
+    }
+  };
+
+  transitionFromKeyValStore = async (sessionId?: number) => {
+    const keyValDb = await keyValDatabaseExists();
+    if (!keyValDb) {
+      return;
+    }
+
+    const transitionCurrentSessionSequences = async (numericSessionId: number, sessionStore: IDBStoreSession) => {
+      const currentSessionSequences = sessionStore.sessionSequences;
+      await Promise.all(
+        Object.keys(currentSessionSequences).map(async (sequenceId) => {
+          const numericSequenceId = parseInt(sequenceId, 10);
+          if (numericSequenceId === sessionStore.currentSequenceId) {
+            await Promise.all(
+              currentSessionSequences[sessionStore.currentSequenceId].events.map(async (event) => {
+                await this.addEventToCurrentSequence(numericSessionId, event);
+              }),
+            );
+          } else if (currentSessionSequences[numericSequenceId].status !== RecordingStatus.SENT) {
+            await this.storeSendingEvents(numericSessionId, currentSessionSequences[numericSequenceId].events);
+          }
+        }),
+      );
```

Comment:
It may be hard to get a sense of the upper bound of promises this may create, but it may be better to batch these. This is just in case this causes too many promises to fire.

Reply:
Implementing it in a batchable way could also make it more readable... this code got so nested